File: back/src/main/java/com/back/global/exception/ErrorCode.java
package com.back.global.exception;

import lombok.Getter;
import lombok.RequiredArgsConstructor;


@Getter
@RequiredArgsConstructor
public enum ErrorCode {
    // Auth
    AUTH_ID_NOT_FOUND("401-1", "존재하지 않는 아이디입니다."),
    AUTH_PASSWORD_MISMATCH("401-2", "비밀번호가 일치하지 않습니다."),
    AUTH_INVALID_BEARER_HEADER("401-3", "Authorization 헤더가 Bearer 형식이 아닙니다."),
    AUTH_INVALID_CREDENTIALS("401-4", "인증 정보가 유효하지 않습니다."),
    AUTH_INVALID_REFRESH_TOKEN("401-5", "유효하지 않은 리프레시 토큰입니다."),
    AUTH_REFRESH_TOKEN_NOT_FOUND_IN_REDIS("401-6", "Redis에 리프레시 토큰이 존재하지 않습니다."),
    AUTH_REFRESH_TOKEN_MISMATCH("401-7", "리프레시 토큰이 일치하지 않습니다."),
    AUTH_LOGIN_REQUIRED("401-8", "로그인 후 이용해주세요."),
    AUTH_FORBIDDEN("403-1", "권한이 없습니다."),

    // User
    USER_NOT_FOUND("404-1", "회원이 존재하지 않습니다."),
    USER_ID_ALREADY_EXISTS("409-1", "이미 사용 중인 아이디입니다."),
    USER_EMAIL_ALREADY_EXISTS("409-2", "이미 사용 중인 이메일입니다."),
    USER_ACCESS_DENIED("403-2", "본인의 정보만 조회할 수 있습니다."),
    USER_NOT_FOUND_OR_DELETED("404-2", "존재하지 않거나 이미 탈퇴한 회원입니다."),

    // Concert
    CONCERT_NOT_FOUND("404-3", "존재하지 않는 콘서트입니다."),

    // Schedule
    CONCERT_SCHEDULE_EMPTY("404-4", "등록된 회차가 없습니다."),
    SCHEDULE_NOT_FOUND("404-5", "존재하지 않는 회차입니다."),
    INVALID_CONCERT_SCHEDULE("400-1", "해당 콘서트의 회차가 아닙니다."),

    // Seat
    SEAT_NOT_FOUND("404-7", "존재하지 않는 좌석입니다."),
    SEAT_SOLD_OUT("400-2", "이미 매진된 좌석입니다."),
    SEAT_ALREADY_SOLD("409-3", "이미 판매 완료된 좌석입니다."),
    SEAT_ALREADY_HOLD("409-4", "이미 점유중인 좌석입니다."),
    SEAT_HELD_BY_OTHER_USER("409-5", "다른 사용자가 선택 중인 좌석입니다."),

    // Ticket
    TICKET_NOT_FOUND_FOR_USER("404-6", "해당 유저의 티켓이 존재하지 않습니다."),
    TICKET_ALREADY_CANCELLED("400-3", "이미 취소된 티켓입니다.");


    private final String resultCode;
    private final String message;

    public int getStatusCode() {
        return Integer.parseInt(resultCode.split("-", 2)[0]);
    }

}

File: back/src/main/kotlin/com/back/global/exception/ErrorCode.kt
package com.back.global.exception

enum class ErrorCode(
    val resultCode: String,
    val message: String
) {
    // global
    BAD_REQUEST("400-4", "잘못된 요청입니다."),

    // Auth
    AUTH_LOGIN_FAILED("401-1", "아이디 또는 비밀번호가 일치하지 않습니다."),
    AUTH_PASSWORD_MISMATCH("401-2", "비밀번호가 일치하지 않습니다."),
    AUTH_INVALID_BEARER_HEADER("401-3", "Authorization 헤더가 Bearer 형식이 아닙니다."),
    AUTH_INVALID_REFRESH_TOKEN("401-5", "유효하지 않은 리프레시 토큰입니다."),
    AUTH_REFRESH_TOKEN_MISMATCH("401-7", "리프레시 토큰이 일치하지 않습니다."),
    AUTH_LOGIN_REQUIRED("401-8", "로그인 후 이용해주세요."),
    AUTH_EXPIRED_ACCESS_TOKEN("401-9", "Access Token이 만료되었습니다."),
    AUTH_INVALID_ACCESS_TOKEN("401-10", "유효하지 않은 Access Token입니다."),
    AUTH_FORBIDDEN("403-1", "권한이 없습니다."),
    AUTH_EMAIL_VERIFICATION_INVALID("400-10", "이메일 인증번호가 올바르지 않거나 만료되었습니다."),
    AUTH_EMAIL_VERIFICATION_REQUIRED("401-12", "이메일 인증이 필요합니다."),
    AUTH_EMAIL_VERIFICATION_TOO_MANY_ATTEMPTS("429-2", "이메일 인증 시도 횟수를 초과했습니다."),
    AUTH_EMAIL_VERIFICATION_RESEND_NOT_ALLOWED("429-3", "이메일 인증번호 재전송 대기 시간이 지나지 않았습니다."),
    AUTH_EMAIL_SEND_FAILED("500-3", "인증 이메일 전송에 실패했습니다."),
    AUTH_REFRESH_TOKEN_ROTATION_FAILED("500-1", "리프레시 토큰 교체 처리 중 오류가 발생했습니다."),

    // User
    USER_NOT_FOUND("404-1", "회원이 존재하지 않습니다."),
    USER_ID_ALREADY_EXISTS("409-1", "이미 사용 중인 아이디입니다."),
    USER_EMAIL_ALREADY_EXISTS("409-2", "이미 사용 중인 이메일입니다."),
    USER_NOT_FOUND_OR_DELETED("404-2", "존재하지 않거나 이미 탈퇴한 회원입니다."),
    USER_NAME_INVALID("400-6", "이름에 공백을 포함할 수 없습니다."),
    FILE_EMPTY("400-7", "업로드할 파일이 비어있습니다."),
    FILE_TOO_LARGE("400-8", "파일 크기가 5MB를 초과합니다."),
    FILE_INVALID_TYPE("400-10", "허용되지 않는 파일 형식입니다. (jpg, jpeg, png, webp만 가능)"),

    // Concert
    CONCERT_NOT_FOUND("404-3", "존재하지 않는 콘서트입니다."),

    // Schedule
    CONCERT_SCHEDULE_EMPTY("404-4", "등록된 회차가 없습니다."),
    INVALID_CONCERT_SCHEDULE("400-1", "해당 콘서트의 회차가 아닙니다."),
    CONCERT_NOT_FOUND_OR_MISMATCH("400-4", "콘서트 정보가 없거나 일치하는 콘서트가 아닙니다."),
    EXPIRED_BOOKING_DEADLINE("400-9", "해당 회차의 예매 가능 시간이 경과되었습니다."),

    // Seat
    SEAT_NOT_FOUND("404-7", "존재하지 않는 좌석입니다."),
    SEAT_ALREADY_SOLD("409-3", "이미 판매 완료된 좌석입니다."),
    SEAT_HELD_BY_OTHER_USER("409-5", "다른 사용자가 선택 중인 좌석입니다."),
    SEAT_HOLD_EXPIRED("409-6", "좌석 점유가 만료되었습니다."),
    INVALID_OCCUPY_TOKEN("409-7", "유효하지 않은 점유 토큰입니다."),

    // Ticket
    TICKET_NOT_FOUND("404-5", "존재하지 않는 티켓입니다."),
    TICKET_NOT_FOUND_FOR_USER("404-6", "해당 유저의 티켓이 존재하지 않습니다."),
    TICKET_ALREADY_CANCELLED("400-3", "이미 취소된 티켓입니다."),
    EXCEED_TICKET_LIMIT("400-2", "회차당 최대 3매까지 예매 가능합니다."),

    // Review
    REVIEW_NOT_FOUND("404-9", "존재하지 않는 리뷰입니다."),
    REVIEW_FORBIDDEN("403-2", "리뷰를 수정/삭제할 권한이 없습니다."),
    REVIEW_NOT_ELIGIBLE("403-4", "해당 콘서트에 대한 유효한 티켓이 없어 리뷰를 작성할 수 없습니다."),
    REVIEW_PERIOD_EXPIRED("403-5", "리뷰 작성 가능 기간이 지났습니다. (콘서트 종료 후 6개월 이내)"),
    REVIEW_ALREADY_EXISTS("409-4", "이미 해당 콘서트에 리뷰를 작성했습니다."),

    // bucket4j
    TOO_MANY_REQUESTS("429-1", "요청이 너무 많습니다."),

    // QUEUE
    QUEUE_TOKEN_NOT_FOUND("401-11", "대기열 접속 정보가 올바르지 않습니다."),
    QUEUE_SESSION_EXPIRED("403-3", "대기열 접속 시간이 만료되었습니다."),

    // WAITING
    WAITING_QUEUE_NOT_FOUND("404-8", "대기열에 등록되지 않은 사용자입니다."),
    WAITING_QUEUE_REGISTER_FAILED("500-2", "대기열 등록 처리 중 오류가 발생했습니다."),
    CONCERT_SOLD_OUT("400-5", "콘서트가 매진되어 대기열이 종료되었습니다.");

    val statusCode: Int
        get() = resultCode.split("-", limit = 2)[0].toInt()
}
